source/include/azure_iot_hub_client.h

```diff
@@ -0,0 +1,411 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved. */
+/* SPDX-License-Identifier: MIT */
+
+/**
+ * @file azure_iot_hub_client.h
+ *
+ * @brief The middleware IoT Hub Client used to connect a device to Azure IoT Hub.
+ *
+ * @note You MUST NOT use any symbols (macros, functions, structures, enums, etc.)
+ * prefixed with an underscore ('_') directly in your application code. These symbols
+ * are part of Azure SDK's internal implementation; we do not document these symbols
+ * and they are subject to change in future versions of the SDK which would break your code.
+ *
+ */
+
+#ifndef AZURE_IOT_HUB_CLIENT_H
+#define AZURE_IOT_HUB_CLIENT_H
+
+#include "azure_iot.h"
+
+#include "azure_iot_mqtt_port.h"
```

Comment:
Please remove this header, since this currently #include core_mqtt.h which means that this gets sucked into user's namespace.  And I don't think it's critical in any case.

Reply:
This port defines a struct AzureIoTMQTT_t, that is required in the struct of AzureIoTHubClient:_internal:xMQTTContext;

Reply:
Is there anyway to get it so that core_mqtt.h doesn't come through to the application namespace?  This is my main concern with the referencing the port.

Reply:
Without using malloc, its bit difficult as we need memory to be allocated by user and pass to us. 

Right now the way we are doing it is making MQTT context type  be part of IoTHub control block and user let user own the IoTHub control block. To make Mqtt context type part of IoTHub control block, we need to know the type definition, which is part of the port (mqtt_port). This is something similar to FreeRTOS port.h/portmacro.h https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/50a23218381574f2ba314109691e1de4f8456684/portable/GCC/ARM_CM4F/portmacro.h#L47  that maps platform specific types and code takes dependency on them.

Reply:
Ahh - that makes sense.  We (or really the compiler) needs to know the sizeof() all these structures - even if they're massively abstracted from the application space - so an app could do static of this structure.

Agree the no malloc() is high-order bit, and any possible solution to this that doesn't involve just giving the compiler the types it needs to do its job feels like it's going to be inherently unstable or do profoundly gross (and unstabel) things with the preprocossor.

I'll noodle a bit more on this, but given our constraints I think what you're doing makes sense here.

Reply:
I'm going to go ahead and mark the conversation of "remove coreMQTT from sneaking out" since I haven't thought of a better way to do this in no-malloc world.  Will let you know if inspiration strikes.